compiler/passes/denormalize.cpp

```diff
@@ -669,3 +676,259 @@ inline bool unsafeExprInBetween(Expr* e1, Expr* e2, Expr* exprToMove,
+         isGoodRefArg(fn, info, true, fn->formals.tail->prev->prev)) )
+    return false;
+
+  if (info.refArg->type->symbol->hasFlag(FLAG_STAR_TUPLE))
+    return false; // codegen requires star tuples to be passed by ref
+
+  for_SymbolSymExprs(refSE, info.refArg) {
+    if (CallExpr* call = toCallExpr(refSE->parentExpr)) {
+      if (call->isPrimitive(PRIM_NO_ALIAS_SET))
+        continue; // any of these are ok
+      if (! call->isPrimitive(PRIM_ASSIGN))
+        return false; // need more work to handle, ex., PRIM_SET_MEMBER
+      // transformRetTempDef() will assert that we have only one PRIM_ASSIGN
+    }
+  }
+
+  return true;
+}
+
+static void transformRetTempDef(DefInfoRR& info) {
+  SymExpr* refUse = nullptr;
+  for_SymbolSymExprs(refSE, info.refArg) {
+    if (CallExpr* call = toCallExpr(refSE->parentExpr))
+      if (call->isPrimitive(PRIM_NO_ALIAS_SET)) {
+        if (call->numActuals() == 1) call->remove();
+        else refSE->remove();
+        continue;
+      }
+    INT_ASSERT(refUse == nullptr); // expect only a single SE
+    refUse = refSE;
+  }
+
+  CallExpr* assignCall = toCallExpr(refUse->parentExpr);
+  Expr* refValueExpr = assignCall->get(2)->remove();
+  INT_ASSERT(isSymExpr(refValueExpr)); // ensure it can be used in 'return'
+  // Do we need to add FLAG_RVV to refValueExpr->symbol()?
+  // At this point temps with FLAG_RVV may occur anywhere due to inlining.
+  assignCall->remove();
+
+  CallExpr* returnCall = toCallExpr(info.fn->body->body.tail);
+  INT_ASSERT(returnCall->isPrimitive(PRIM_RETURN));
+  INT_ASSERT(toSymExpr(returnCall->get(1))->symbol() == gVoid);
+  returnCall->get(1)->replace(refValueExpr);
+
+  info.refArg->defPoint->remove();
+  info.fn->retType = info.refArg->type;
+  info.fn->removeFlag(FLAG_FN_RETARG);
+}
+
+///////////
+//
+// transformRetTempUse() will replace:
+//   call fn(args, ret_tmp)
+// with:
+//   move(ret_tmp, call fn(args))
+//
+// collapseTrivialMoves() will reduce to the original 'call_temp = fn(...)'
+
+class UseInfoRR {
+public:             // in 'call fn(args, ret_tmp)' above, the SymExprs for:
+  SymExpr* fnSE;    //  'fn'
+  SymExpr* tempSE;  //  'ret_tmp'
+};
+
+// Return true if the desired pattern is present.
+static bool acceptableUse(DefInfoRR& defInfo, SymExpr* fnUse,
+                          UseInfoRR& useInfo) {
+  CallExpr* call = toCallExpr(fnUse->parentExpr);
+  if (call == nullptr || call->resolvedFunction() != defInfo.fn)
+    return false;
+
+  INT_ASSERT(call == call->getStmtExpr());
+  SymExpr* tempSE = toSymExpr(defInfo.skipLnFnArgs ?
+                      call->argList.tail->prev->prev : call->argList.tail);
+  // the temp is usually called "ret_tmp", however not necessarily
+  INT_ASSERT(tempSE->symbol()->type == defInfo.refArg->type); // fyi
+
+  useInfo = { fnUse, tempSE };
+  return true;
+}
+
+static void transformRetTempUse(UseInfoRR& info) {
+  Expr*   fnCall  = info.fnSE->parentExpr;
+  Symbol* retTemp = info.tempSE->symbol();
+  //fyi:
```

Comment:
Remove the `fyi`, I think the asserts speak for themselves? If they're arcane you already have a comment.